<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [ ]:
import pandas as pd
import psycopg2

In [ ]:
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432 

In [ ]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

n = 10

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [ ]:
query_3_1 = f'''select count(*)
           from public.vacancies
        '''

In [ ]:
df = pd.read_sql_query(query_3_1, connection)
df

/var/folders/xx/g9ddp_l94kn2b3bnr9fpjm3w0000gn/T/ipykernel_8725/3851344990.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [ ]:
query_3_2 = f'''select count(*)
           from public.employers
        '''

In [ ]:
df_3_2 = pd.read_sql_query(query_3_2, connection)
df_3_2

/var/folders/xx/g9ddp_l94kn2b3bnr9fpjm3w0000gn/T/ipykernel_8725/2741247974.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_2 = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [ ]:
query_3_3 = f'''select count(*)
           from public.areas
        '''

In [ ]:
df_3_3= pd.read_sql_query(query_3_3, connection)
df_3_3

/var/folders/xx/g9ddp_l94kn2b3bnr9fpjm3w0000gn/T/ipykernel_8725/3776500924.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_3= pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [ ]:
query_3_4 = f'''select count(*)
           from public.industries
        '''

In [ ]:
df_3_4 = pd.read_sql_query(query_3_4, connection)
df_3_4

/var/folders/xx/g9ddp_l94kn2b3bnr9fpjm3w0000gn/T/ipykernel_8725/764330752.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_4 = pd.read_sql_query(query_3_4, connection)


count
0    294

***

*Из предварительного анализа данных видно, что в базе HH опубликовано более 49 тысяч вакансий. Также подсчитано количество количество различных работодатей (23.5 тысячи), регионов (1362) и сфер деятельности компаний (294) из соответствующих таблиц. Стоит отметить, что в опубликованных вакансиях необязательно будут присутствовать все реботодатели, регионы и сферы деятельности. Например, дальше в задании 5.2 мы увидем, что в 593 из 1362 регионов нет вакансий.*

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [ ]:
query_4_1 = f'''select A.name area, count(*) cnt
           from public.vacancies V join public.areas A on V.area_id = A.id
           group by area
           order by cnt DESC
           limit {n}
        '''

In [ ]:
df_4_1 = pd.read_sql_query(query_4_1, connection)
df_4_1

/var/folders/xx/g9ddp_l94kn2b3bnr9fpjm3w0000gn/T/ipykernel_8725/1522353868.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_1 = pd.read_sql_query(query_4_1, connection)


area   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892
5     Екатеринбург  1698
6  Нижний Новгород  1670
7           Казань  1415
8        Краснодар  1301
9           Самара  1144

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [ ]:
query_4_2 = f'''select count(*)
           from public.vacancies 
           where (salary_from IS NOT NULL) OR (salary_to IS NOT NULL)
        '''

In [ ]:
df_4_2 = pd.read_sql_query(query_4_2, connection)
df_4_2

/var/folders/xx/g9ddp_l94kn2b3bnr9fpjm3w0000gn/T/ipykernel_8725/3761913292.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_2 = pd.read_sql_query(query_4_2, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [ ]:
query_4_3 = f'''select AVG(salary_from)
           from public.vacancies 
        '''
query_4_3_2 = f'''select AVG(salary_to)
           from public.vacancies 
        '''

In [ ]:
df_4_3 = pd.read_sql_query(query_4_3, connection)
display(df_4_3)

df_4_3_2 = pd.read_sql_query(query_4_3_2, connection)
display(df_4_3_2)

/var/folders/xx/g9ddp_l94kn2b3bnr9fpjm3w0000gn/T/ipykernel_8725/451436104.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_3 = pd.read_sql_query(query_4_3, connection)


avg
0  71064.657901

/var/folders/xx/g9ddp_l94kn2b3bnr9fpjm3w0000gn/T/ipykernel_8725/451436104.py:4: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_3_2 = pd.read_sql_query(query_4_3_2, connection)


avg
0  110536.741923

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [ ]:
query_4_4 = f'''select schedule, employment, count(*) cnt
           from public.vacancies 
           group by schedule, employment
           order by cnt DESC
        '''

In [ ]:
df_4_4 = pd.read_sql_query(query_4_4, connection)
df_4_4

/var/folders/xx/g9ddp_l94kn2b3bnr9fpjm3w0000gn/T/ipykernel_8725/1636312543.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_4 = pd.read_sql_query(query_4_4, connection)


schedule           employment    cnt
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [ ]:
query_4_5 = f'''select experience, count(*)
           from public.vacancies 
           group by experience
           order by count(*)
        '''

In [ ]:
df_4_5 = pd.read_sql_query(query_4_5, connection)
df_4_5

/var/folders/xx/g9ddp_l94kn2b3bnr9fpjm3w0000gn/T/ipykernel_8725/802873672.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_5 = pd.read_sql_query(query_4_5, connection)


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

*Из детального анализа данных видно, что в базе HH размещены вакансии не только в регионах России, но и по всему миру. Причем количество вакансий в Минске и Алма-Ате входит в топ-10 регионов по количеству вакансий. Самое большое количество вакансий, естественно, в Москве и Санкт-Петербурге.* 
*Также можно видеть, что чуть меньше, чем в половине вакансий указана нижняя или верхняя граница зарплатной вилки, и средняя зарплата по всем вакансиям колеблется от 71 до 110 тысяч, что выглядит довольно высокой ЗП для России.*


*Также интересно, что работодатели чаще всего ищут сотрудников с опытом работы от 1 до 3 лет - таких вакнсий больше половины (более 26 тысяч). Возможно, это связано с тем, что работодателю достаточно навыков, которые приобрел сотрудник за такое время работы. Около 15% вакансий опубликовано для сотрудников без опыта работы*
*Самым популярным типом трудоустройства является полная занятость (около 45 тысяч вакансий), и только около 3-4 тысяч вакансий имеют тип "стажировка" или "частичная занятость". Чаще всего работодатели ищут сотрудников на полный день, но и удаленная работа не является редкостью*

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [ ]:
query_5_1 = f'''select E.name employer, count(*) cnt
           from public.vacancies V join public.employers E on V.employer_id = E.id
           group by employer
           order by cnt DESC
           limit {n}
        '''

In [ ]:
df_5_1 = pd.read_sql_query(query_5_1, connection)
df_5_1

/var/folders/xx/g9ddp_l94kn2b3bnr9fpjm3w0000gn/T/ipykernel_8725/556590240.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_1 = pd.read_sql_query(query_5_1, connection)


employer   cnt
0                 Яндекс  1933
1             Ростелеком   491
2               Тинькофф   444
3                   СБЕР   428
4          Газпром нефть   331
5            ИК СИБИНТЕК   327
6                    МТС   292
7                DataArt   247
8  Совкомбанк Технологии   204
9             Первый Бит   176

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [ ]:
query_5_2 = f'''select A.name, count(DISTINCT E.id) emp_cnt, count(V.id) vcn_cnt
           from public.areas A 
            left join public.employers E on E.area = A.id 
            left join public.vacancies V on V.area_id = A.id
           where V.id IS NULL
           group by A.id
           order by emp_cnt DESC
           
        '''

In [ ]:
df_5_2 = pd.read_sql_query(query_5_2, connection)
df_5_2

/var/folders/xx/g9ddp_l94kn2b3bnr9fpjm3w0000gn/T/ipykernel_8725/762398461.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_2 = pd.read_sql_query(query_5_2, connection)


name  emp_cnt  vcn_cnt
0                Россия      410        0
1             Казахстан      207        0
2    Московская область       75        0
3    Краснодарский край       19        0
4    Ростовская область       18        0
..                  ...      ...      ...
588             Скидель        0        0
589          Белоозерск        0        0
590           Орловский        0        0
591          Павловская        0        0
592             Таксимо        0        0

[593 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [ ]:
query_5_3 = f'''select E.name, count(DISTINCT V.area_id)
           from public.vacancies V join public.employers E on V.employer_id = E.id
           group by E.name
           order by count(V.area_id) DESC
           limit {n}
        '''

In [ ]:
df_5_3 = pd.read_sql_query(query_5_3, connection)
df_5_3

/var/folders/xx/g9ddp_l94kn2b3bnr9fpjm3w0000gn/T/ipykernel_8725/3556584873.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_3 = pd.read_sql_query(query_5_3, connection)


name  count
0                 Яндекс    181
1             Ростелеком    152
2               Тинькофф     43
3                   СБЕР     24
4          Газпром нефть     24
5            ИК СИБИНТЕК     46
6                    МТС     55
7                DataArt      2
8  Совкомбанк Технологии     26
9             Первый Бит     43

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [ ]:
query_5_4 = f'''select count(*)
           from  public.employers E left join public.employers_industries EI on E.id = EI.employer_id
            where industry_id IS NULL
        '''

In [ ]:
df_5_4 = pd.read_sql_query(query_5_4, connection)
df_5_4

/var/folders/xx/g9ddp_l94kn2b3bnr9fpjm3w0000gn/T/ipykernel_8725/2345892774.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_4 = pd.read_sql_query(query_5_4, connection)


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [ ]:
query_5_5 = f'''select E.name 
           from public.employers E 
            join public.employers_industries EI on E.id = EI.employer_id 
            group by E.name
            having count(*) = 4
            order by E.name
        '''

In [ ]:
df_5_5 = pd.read_sql_query(query_5_5, connection)
df_5_5

/var/folders/xx/g9ddp_l94kn2b3bnr9fpjm3w0000gn/T/ipykernel_8725/480638788.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_5 = pd.read_sql_query(query_5_5, connection)


name
0        101 Интернет
1            21vek.by
2                2ГИС
3                  2К
4         4 пикселя +
...               ...
1131            ЮРИОН
1132  ЮТИП Технологии
1133           ЯКласс
1134            ЯрНео
1135    Яшин&Партнёры

[1136 rows x 1 columns]

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [ ]:
query_5_6 = f'''select count(*)
           from public.employers E 
            join public.employers_industries EI on E.id = EI.employer_id 
            join public.industries I on I.id = EI.industry_id
           where  I.name = 'Разработка программного обеспечения'
        '''

In [ ]:
df_5_6 = pd.read_sql_query(query_5_6, connection)
df_5_6

/var/folders/xx/g9ddp_l94kn2b3bnr9fpjm3w0000gn/T/ipykernel_8725/3865320243.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_6 = pd.read_sql_query(query_5_6, connection)


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [ ]:
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8'
cities = pd.read_html(url, index_col=0)[1]
cities = tuple(cities.iloc[:, 0])
cities

('Москва',
 'Санкт-Петербург',
 'Новосибирск',
 'Екатеринбург',
 'Казань',
 'Нижний Новгород',
 'Красноярск',
 'Челябинск',
 'Самара',
 'Уфа',
 'Ростов-на-Дону',
 'Краснодар',
 'Омск',
 'Воронеж',
 'Пермь',
 'Волгоград')

In [ ]:
query_5_7 = f'''(select A.name, count(*)
           from public.vacancies V
           left join public.employers E on V.employer_id = E.id
           left join public.areas A on V.area_id = A.id
           where  (E.name = 'Яндекс') and (A.name in {cities})
           group by A.name
           order by 2)
           
             union all
             
           (select 'Total', count(*)
           from public.vacancies V
           left join public.employers E on V.employer_id = E.id
           left join public.areas A on V.area_id = A.id
           where  (E.name = 'Яндекс') and (A.name in {cities}))
        '''

In [ ]:
df_5_7 = pd.read_sql_query(query_5_7, connection)
df_5_7

/var/folders/xx/g9ddp_l94kn2b3bnr9fpjm3w0000gn/T/ipykernel_8725/43835236.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_7 = pd.read_sql_query(query_5_7, connection)


name  count
0             Total    485
1              Омск     21
2         Челябинск     22
3        Красноярск     23
4         Волгоград     24
5             Пермь     25
6            Казань     25
7    Ростов-на-Дону     25
8               Уфа     26
9            Самара     26
10        Краснодар     30
11          Воронеж     32
12      Новосибирск     35
13  Нижний Новгород     36
14     Екатеринбург     39
15  Санкт-Петербург     42
16           Москва     54

*Из анализа работодателей видно, что основными компаниями "гигантами", которые размещают больше всего вакансий в различных регионах, являются Яндекс, Ростелеком, Тинькофф, СБЕР и Газпром нефть. Удивительно, что Тинькофф превосходит СБЕР по количеству вакансий и регионов вакансий, хотя он гораздо меньше СБЕРа. Можно предположить, что это связано со спецификой бизнеса Банка, который не имеет оффлайн-офисов и нанимает сотрудников-операторов колл-центра на удаленную работу по всей России. Большое количество вакансий Яндекса в различных регионах можно объяснить тем, что в Яндексе есть такие направления бизнеса, как, наприммер, Яндекс-такси, яндекс-маркет, яндекс-еда, которые генерят большое количество вакансий для таксистов и курьеров и покрывают территориально большое количество регионов РФ и за пределами РФ. Ростелеком также нанимает большое количество сотрудников, тк является главным провайдером услуг доступа в Интернет, телевидения и телефонной связи по территории всей РФ.*

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [ ]:
query_6_1 = f'''select count(*)
           from public.vacancies
           where (lower(name) like '%data%') or (lower(name) like '%данн%')
        '''

In [ ]:
df_6_1 = pd.read_sql_query(query_6_1, connection)
df_6_1

/var/folders/xx/g9ddp_l94kn2b3bnr9fpjm3w0000gn/T/ipykernel_8725/1300834887.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_1 = pd.read_sql_query(query_6_1, connection)


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [ ]:
query_6_2 = f'''select count(*)
           from public.vacancies
           where ((lower(name) like '%исследователь данных%') or (lower(name)  like '%data scientist%') or (lower(name) like '%data science%') or ((name like '%ML%') and (name not like '%HTML%')) 
            or (lower(name) like '%machine learning%') or (lower(name) like '%машинн%обучен%')) and 
            ((experience = 'Нет опыта') or (lower(name) like '%junior%') or (employment = 'Стажировка'))
        '''

In [ ]:
df_6_2 = pd.read_sql_query(query_6_2, connection)
df_6_2

/var/folders/xx/g9ddp_l94kn2b3bnr9fpjm3w0000gn/T/ipykernel_8725/2048769617.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_2 = pd.read_sql_query(query_6_2, connection)


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [ ]:
query_6_3 = f'''select count(*)
           from public.vacancies
           where ((lower(name) like '%исследователь данных%') or (lower(name)  like '%data scientist%') or (lower(name) like '%data science%') or ((name like '%ML%') and (name not like '%HTML%')) 
            or (lower(name) like '%machine learning%') or (lower(name) like '%машинн%обучен%')) and 
           ((key_skills like '%SQL%') or (lower(key_skills) like '%postgres%'))
        '''

In [ ]:
df_6_3 = pd.read_sql_query(query_6_3, connection)
df_6_3

/var/folders/xx/g9ddp_l94kn2b3bnr9fpjm3w0000gn/T/ipykernel_8725/3192746363.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_3 = pd.read_sql_query(query_6_3, connection)


count
0    201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [ ]:
query_6_4 = f'''select count(*)
           from public.vacancies
           where ((lower(name) like '%исследователь данных%') or (lower(name)  like '%data scientist%') or (lower(name) like '%data science%') or ((name like '%ML%') and (name not like '%HTML%')) 
            or (lower(name) like '%machine learning%') or (lower(name) like '%машинн%обучен%')) and 
           (lower(key_skills) like '%python%')
        '''

In [ ]:
df_6_4 = pd.read_sql_query(query_6_4, connection)
df_6_4

/var/folders/xx/g9ddp_l94kn2b3bnr9fpjm3w0000gn/T/ipykernel_8725/1319374162.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_4 = pd.read_sql_query(query_6_4, connection)


count
0    351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [ ]:
query_6_5 = f'''select round(AVG(length(key_skills) - length(replace(key_skills, CHR(9), '')) + 1), 2)
           from public.vacancies
           where ((lower(name) like '%исследователь данных%') or (lower(name)  like '%data scientist%') or (lower(name) like '%data science%') or ((name like '%ML%') and (name not like '%HTML%')) 
            or (lower(name) like '%machine learning%') or (lower(name) like '%машинн%обучен%')) 
            and (key_skills is not null)
        '''

In [ ]:
df_6_5 = pd.read_sql_query(query_6_5, connection)
df_6_5

/var/folders/xx/g9ddp_l94kn2b3bnr9fpjm3w0000gn/T/ipykernel_8725/1566661722.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_5 = pd.read_sql_query(query_6_5, connection)


round
0   6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [ ]:
query_6_6 = f'''select experience, round(avg(coalesce((salary_from+salary_to)/2, salary_from, salary_to))) avg_salary
           from public.vacancies
           where ((lower(name) like '%исследователь данных%') or (lower(name)  like '%data scientist%') or (lower(name) like '%data science%') or ((name like '%ML%') and (name not like '%HTML%')) 
            or (lower(name) like '%machine learning%') or (lower(name) like '%машинн%обучен%')) 
             and (salary_from is not null or salary_to is not null)   
          group by experience     
        '''

In [ ]:
df_6_6 = pd.read_sql_query(query_6_6, connection)
df_6_6

/var/folders/xx/g9ddp_l94kn2b3bnr9fpjm3w0000gn/T/ipykernel_8725/158571054.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_6 = pd.read_sql_query(query_6_6, connection)


experience  avg_salary
0           Нет опыта     74643.0
1  От 1 года до 3 лет    139675.0
2       От 3 до 6 лет    243115.0

***

7. Найдем общее количество вакансий для Дата-Сайентистов:

In [ ]:
query_6_7 = f'''select count(*)
           from public.vacancies
           where ((lower(name) like '%исследователь данных%') or (lower(name)  like '%data scientist%') or (lower(name) like '%data science%') or ((name like '%ML%') and (name not like '%HTML%')) 
            or (lower(name) like '%machine learning%') or (lower(name) like '%машинн%обучен%')) 
        '''
df_6_7 = pd.read_sql_query(query_6_7, connection)
df_6_7

/var/folders/xx/g9ddp_l94kn2b3bnr9fpjm3w0000gn/T/ipykernel_8725/694533841.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_7 = pd.read_sql_query(query_6_7, connection)


count
0    480

*Предметный анализ показал, что в базе размещено около 480 вакансий для дата-сайнтистов, из них 51 (менее 10%) на Junior-позицию. На входе дата-сайнтисту без опыта работы работодатели готовы платить около 75 тысяч. Средняя заработная плата для специалистов без опыта и с опытом 1-3 года отличается почти в 2 раза. В требуемых навыках Python указан в 351 вакансии, что неудивительно, тк Python  - главный инстумент DS.  А вот SQL или postgres менее важны для работодателей: навык фигурирует в менее чем половине вакансий (201 вакансия)*

# Общий вывод по проекту

### Проведем дополнительные исследования

1. Напишем запрос, который выведет нижнюю вилку заработной платы для каждого региона (для вакансий, в которых заполнено поле salary_from). Также  выведем количество вакансий, которые,  участвовали в расчете среднего.

In [ ]:
query_7_1 = f'''select A.name area, AVG(salary_from), AVG(salary_to), count(*)
           from public.vacancies V join public.areas A on V.area_id = A.id
           where salary_from is not NULL
           group by area
           order by AVG(salary_from) DESC
           limit {40}
        '''
df_7_1 = pd.read_sql_query(query_7_1, connection)
df_7_1

/var/folders/xx/g9ddp_l94kn2b3bnr9fpjm3w0000gn/T/ipykernel_8725/3642157989.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_7_1 = pd.read_sql_query(query_7_1, connection)


area            avg            avg  count
0                            Германия  348270.000000  557232.000000      2
1                             Испания  300000.000000            NaN      1
2                          Черногория  282501.000000  331208.000000      3
3                               Литва  263018.500000            NaN      2
4                             Кишинёв  233794.000000  292242.000000      1
5                         Таджикистан  233794.000000  292242.000000      1
6                                Кипр  210058.054054  272206.058824     37
7                              Турция  201522.235294  263917.384615     17
8                              Сербия  200802.909091  277630.500000     11
9      Заполярный (Ямало-Ненецкий АО)  200000.000000  200000.000000      1
10                           Болгария  200000.000000  300000.000000      1
11                            Армения  181509.449275  242843.513514     69
12                                США  178808.333333  208678.666667      6
13                            Эстония  175346.000000            NaN      1
14                             Польша  175346.000000            NaN      1
15                             Латвия  175346.000000            NaN      1
16                              Чехия  174037.600000  167170.000000      5
17                            Тбилиси  173277.116667  249599.978723     60
18               Дзержинск (Беларусь)  171837.666667  268860.333333      3
19                     Великобритания  167170.000000  250754.500000      2
20                            Зарайск  160000.000000            NaN      1
21                            Кутаиси  157155.666667  350691.000000      3
22  Некрасовский (Московская область)  150000.000000  250000.000000      1
23                         Коммунарка  150000.000000            NaN      1
24                     Старая Купавна  150000.000000            NaN      1
25                          Иннополис  142785.714286  177500.000000     14
26                              Можга  140000.000000            NaN      1
27                           Могойтуй  140000.000000            NaN      1
28                          Карымское  140000.000000            NaN      1
29                             Батуми  137296.500000  238442.727273     12
30                             Усинск  130400.000000  182500.000000      4
31                              Ленск  130000.000000  180000.000000      1
32                    Железнодорожный  127875.000000  250000.000000      4
33                            Кинешма  125000.000000  300000.000000      2
34                         Нидерланды  117000.000000  220500.000000      2
35                              Лобня  115555.555556  185833.333333      9
36                          Мичуринск  114000.000000            NaN      1
37                             Москва  113409.296992  158134.521067   1330
38                             Крымск  110000.000000  210000.000000      2
39                               Киев  109591.000000  172423.000000      4

2. Проверим, нет ли выбросов в заработной плате, выведя топ 5 максимальных заработных плат по верхней границе заработной вилки:

In [ ]:
query_7_2 = f'''select salary_to
           from public.vacancies 
           where salary_to is not null
           order by 1 DESC
           limit 5
        '''
df_7_2= pd.read_sql_query(query_7_2, connection)
df_7_2

/var/folders/xx/g9ddp_l94kn2b3bnr9fpjm3w0000gn/T/ipykernel_8725/1579644310.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_7_2= pd.read_sql_query(query_7_2, connection)


salary_to
0    1000000
1    1000000
2    1000000
3     950000
4     935176

3. Проверим, компании какой из сфер деятельности чаще всего размещают вакансии:

In [ ]:
query_7_3 = f'''select I.name, count(*) cnt 
           from public.employers E 
            join public.employers_industries EI on E.id = EI.employer_id 
            join public.industries I on I.id = EI.industry_id
            group by I.name
            order by cnt DESC
            limit 10
        '''
df_7_3 = pd.read_sql_query(query_7_3, connection)
df_7_3

/var/folders/xx/g9ddp_l94kn2b3bnr9fpjm3w0000gn/T/ipykernel_8725/1788727887.py:9: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_7_3 = pd.read_sql_query(query_7_3, connection)


name   cnt
0                Разработка программного обеспечения  3553
1  Системная интеграция,  автоматизации технологи...  2993
2  Интернет-компания (поисковики, платежные систе...  1675
3  Маркетинговые, рекламные, BTL, дизайнерские, E...   798
4                              Консалтинговые услуги   662
5                                 Кадровые агентства   389
6  Электронно-вычислительная, оптическая, контрол...   374
7                       Розничная сеть (продуктовая)   368
8                                   Интернет-магазин   351
9                        Архитектура, проектирование   346

In [ ]:
connection.close()

*Подводя итог, хочется отметить, что в представленной выборке вакансий фигурирует очень большое количество вакансий из сферы ИТ тк из дополнительного исследования видно, что больше всего вакансий размещают компании из сфер деятельности "Разработка программного обеспечения", "Системная интеграция, автоматизации технол...", "Интернет-компания (поисковики, платежные систе...". Если открыть сайт HH.ru, можно увидеть, что на самом сайте больше всего вакансий в сфере "Розничная торговля", и таких вакансий в 2 раза больше чем вакансий в сфере ИТ.* 

*Проведя анализ, мы увидели, что данные в данных содержатся различные данные о конкретных вакансиях, в том числе вакансий для дата-сантистов, а также о различнвх работодателях, отраслях и регионах. Отсюда можно сделать предварительный вывод о том, что эти данные могут подойти для того, чтобы попыться создать на них модель машинного обучения, которая будет рекомендовать вакансии клиентам кадрового агентства, претендующим на позицию Data Scientist*

*Скорее всего именно из-за того, что в данных представлено много вакансий из сферы ИТ, у нас получилась очень высокая средняя заработная плата (71-110 тысяч), ведь по данным Росстата средняя ЗП в России за февраль 2023 года составила около 65 тысяч.*

*В связи с этим возникло желание проверить заработную плату на выбросы в данных. В целом максимальное значение верхней границы зарплатной вилки - 1 млн рублей - не выглядит странным, и таких вакансий всего 3.*

*Тк в базе много вакансий заграницей, встает вопрос о валюте, в которой указана ЗП. В запросе 4.6 выведены регионы с самой высокой нижней границей ЗП, в ТОП-9 нет ни одного региона РФ, то есть самые высокие ЗП предлагают заграницей. Однако количество таких вакансий единично. И по этим единичным вакансиям можно предположить, что ЗП указана в рублях, однако в дальнейшем стоит детальнее проанализировать, в какой валюте указана заработная плата.*
